crates/bevy_ecs/src/entity/map_entities.rs

```diff
@@ -3,46 +3,111 @@ use crate::{
-/// inject the entity mapping strategy between your `MapEntities` type and the current world
-/// (usually by using an [`EntityHashMap<Entity>`] between source entities and entities in the
-/// current world).
+/// This is implemented by default for types that implement [`IterEntities`] or
+/// where `&T` implements [`IntoIterator`].
 ///
-/// Implementing this trait correctly is required for properly loading components
-/// with entity references from scenes.
+/// It may be useful to implement directly for types that can't produce an
+/// iterator for lifetime reasons, such as those involving internal mutexes.
+pub trait MapEntities {
+    /// Apply an operation to all contained entities.
+    fn map_entities<F: FnMut(Entity)>(&self, f: F);
+}
+
+/// Apply an operation to mutable references to all entities in a container.
 ///
-/// ## Example
+/// This is implemented by default for types that implement [`IterEntitiesMut`]
+/// or where `&mut T` implements [`IntoIterator`].
 ///
-/// ```
-/// use bevy_ecs::prelude::*;
-/// use bevy_ecs::entity::MapEntities;
+/// It may be useful to implement directly for types that can't produce an
+/// iterator for lifetime reasons, such as those involving internal mutexes.
 ///
-/// #[derive(Component)]
-/// struct Spring {
-///     a: Entity,
-///     b: Entity,
-/// }
+/// Because the operation receives mutable references, it may alter the
+/// contained entity IDs via a mechanism such as an [`EntityMapper`].
+pub trait MapEntitiesMut {
+    /// Apply an operation to mutable references to all entities.
+    fn map_entities_mut<F: FnMut(&mut Entity)>(&mut self, f: F);
+}
+
+impl<T> MapEntities for T
+where
+    T: IterEntities,
+{
+    fn map_entities<F: FnMut(Entity)>(&self, f: F) {
+        self.iter_entities().for_each(f);
+    }
+}
+
+impl<T> MapEntitiesMut for T
+where
+    T: MapEntities + IterEntitiesMut,
+{
+    fn map_entities_mut<F: FnMut(&mut Entity)>(&mut self, f: F) {
+        self.iter_entities_mut().for_each(f);
+    }
+}
+
+/// Produce an iterator over all contained entities.
 ///
-/// impl MapEntities for Spring {
-///     fn map_entities<M: EntityMapper>(&mut self, entity_mapper: &mut M) {
-///         self.a = entity_mapper.map_entity(self.a);
-///         self.b = entity_mapper.map_entity(self.b);
-///     }
-/// }
-/// ```
-pub trait MapEntities {
-    /// Updates all [`Entity`] references stored inside using `entity_mapper`.
-    ///
-    /// Implementors should look up any and all [`Entity`] values stored within `self` and
-    /// update them to the mapped values via `entity_mapper`.
-    fn map_entities<M: EntityMapper>(&mut self, entity_mapper: &mut M);
+/// This is implemented by default for types  where `&T` implements
+/// [`IntoIterator`].
+///
+/// It may be useful to implement directly for types that can't produce an
+/// iterator for lifetime reasons, such as those involving internal mutexes.
+pub trait IterEntities {
```

Comment:
Do we need 2 separate traits? Maybe we could have a single trait with just two methods?

Reply:
:facepalm: 

Dunno why I thought this wasn't a viable option in my previous comment. Been staring at this for too long I think. Brain broke.

Yep, it's now one trait.